### PR TITLE
[daggy-u][dbt] Update dbt course to use DbtProject

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR updates the dbt course in Dagster University to use DbtProject.

To do outside this PR:
- update the knowledge check/quiz in Lesson 4. The `DAGSTER_DBT_PARSE_PROJECT_ON_LOAD` env var is no longer used.

### CODE REVIEW

**Comment by maximearmstrong** [comment]
* **#23098** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23098?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maximearmstrong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by cmpadden** [comment]
This is looking great.

@maximearmstrong we will also need to update the  "completed" `dagster-and-dbt` project here:

- https://github.com/dagster-io/project-dagster-university/tree/module/dagster-and-dbt

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from pathlib import Path
+
+from dagster_dbt import DbtProject
+```
+
+The `Path` class from the `pathlib` standard library will help us create an accurate pointer to where our dbt project is. The `DbtProject` class is imported from the `dagster_dbt` package that we installed earlier. After the import, add the following code: 
+
+```python
+dbt_project = DbtProject(
+  project_dir=Path(__file__).joinpath("..", "..", "..", "analytics").resolve(),
```
Comment (cmpadden) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md:23]: I know we did this in the previous lesson as well, but I'm not a fan of repeating `..` relative paths.

Would it be more intuitive if we used dagster home, or `DagsterInstance.get().root_directory()`? This can merge before this change, but wanted to bring it up.

Comment (maximearmstrong) [reply docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md:23]: I understand your point, but I think we should avoid this for two main reasons:
- it is common to see a dbt project outside the Dagster project directory - we would need to resolve the path using `DagsterInstance.get().root_directory()` and more `..` to indicate the dbt project location.
- the current scaffold template uses `..` relative paths. If we want to change that in Dagster University, we should start by the scaffold.

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Create a new file in the `assets` directory called `dbt.py`. Open that file and add the following import at the top:
+
+```python
+from pathlib import Path
+
+from dagster_dbt import DbtProject
+```
+
+The `Path` class from the `pathlib` standard library will help us create an accurate pointer to where our dbt project is. The `DbtProject` class is imported from the `dagster_dbt` package that we installed earlier. After the import, add the following code:
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md:19]:
```suggestion
The `Path` class from the `pathlib` standard library will help us create an accurate pointer to where our dbt project is. The `DbtProject` class is imported from the `dagster_dbt` package that we installed earlier.

After the imports, add the following code: 
```

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+title: 'Lesson 3: Representing the dbt project in Dagster'
+module: 'dagster_dbt'
+lesson: '3'
+---
+
+# Representing the dbt project in Dagster
+
+As you’ll frequently point your Dagster code to the `target/manifest.json` file and your dbt project in this course, it’ll be helpful to keep a reusable representation of the dbt project. This can be easily done using the `DbtProject` class.
+
+Create a new file in the `assets` directory called `dbt.py`. Open that file and add the following import at the top:
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md:11]:
```suggestion
In the `assets` directory, create a new `dbt.py` file and add the following imports:
```

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+```python
+dbt_project = DbtProject(
+  project_dir=Path(__file__).joinpath("..", "..", "..", "analytics").resolve(),
+)
+```
+
+This code creates a representation of the dbt project called `dbt_project`. The code defining the location of the project directory might look a little complicated, so let’s break it down:
+
+- It uses the location of the `dbt.py` file (via `__file__`) as a point of reference for finding the dbt project
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md:29]: Nit - what does `it` in `It uses the location of...` refer to here? The code? Something specifically in the code?

File: docs/dagster-university/pages/dagster-dbt/lesson-3/5-loading-dbt-models-into-dagster-as-assets.md [hunk 5]
```diff
@@ -28,32 +28,26 @@ We’ll only create one `@dbt_assets` definition for now, but in a later lesson,
 
 ## Loading the models as assets
 
-1. Create a new file in the `assets` directory called `dbt.py`.
+1. Navigate to the `assets/dbt.py` created earlier and open it.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/5-loading-dbt-models-into-dagster-as-assets.md:31]:
```suggestion
1. Open the `assets/dbt.py` file.
```

Let's just tell them to open the file and keep it simple

File: docs/dagster-university/pages/dagster-dbt/lesson-3/5-loading-dbt-models-into-dagster-as-assets.md [hunk 6]
```diff
@@ -28,32 +28,26 @@ We’ll only create one `@dbt_assets` definition for now, but in a later lesson,
...
-
 4. Now, use the `@dbt_assets` decorator to create a new asset function and provide it with a reference to the manifest:
 
    ```python
    @dbt_assets(
-       manifest=dbt_manifest_path,
+       manifest=dbt_project.manifest_path,
    )
    def dbt_analytics(context: AssetExecutionContext, dbt: DbtCliResource):
    ```
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-3/5-loading-dbt-models-into-dagster-as-assets.md:53]: Step 3 strikes me as a little odd when I look at what Step 4 says. Here, it doesn't look like we're actually instructing the user to do anything - I was a little confused and left wondering what I was supposed to do with the code.

What if we combined Steps 3 and 4 to do something like the following? Feel free to wordsmith:

---

3. Next, we'll use the `@dbt_assets` decorator to create a new asset function and provide it with a reference to the project's manifest file:

   ```python
   @dbt_assets(
       manifest=dbt_project.manifest_path,
   )
   def dbt_analytics(context: AssetExecutionContext, dbt: DbtCliResource):
   ```

   Here, we used `dbt_project.manifest_path` to provide the reference to the project's manifest file. This is possible because the `dbt_project` representation we created earlier contains the manifest path, accessible by using the `manifest_path` attribute.

File: docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md [hunk 7]
```diff
@@ -6,79 +6,30 @@ lesson: '4'
 
 # Speeding up the development cycle
 
-By now, you’ve had to run `dbt parse` and reload your code location quite frequently, which doesn’t feel like the cleanest developer experience.
+By now, you’ve had to run `dbt parse` to create the manifest file and reload your code location quite frequently, which doesn’t feel like the cleanest developer experience.
 
-Before we move on, we’ll reduce the number of steps in the feedback loop. We'll automate the `dbt parse` command by taking advantage of the `DbtCliResource` that we wrote earlier.
+Before we move on, we’ll reduce the number of steps in the feedback loop. We'll automate the creation the manifest file by taking advantage of the `DbtProject` that we wrote earlier.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md:11]:
```suggestion
Before we move on, we’ll reduce the number of steps in the feedback loop. We'll automate the creation of the manifest file by taking advantage of the `DbtProject` that we wrote earlier.
```

Comment (erinkcochran87) [reply docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md:11]: Also, should it be `dbt_project` to be consistent?

File: docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md [hunk 8]
```diff
@@ -6,79 +6,30 @@ lesson: '4'
...
-dbt_manifest_path = (
-    dbt_resource.cli(
-        ["--quiet", "parse"],
-        target_path=Path("target"),
-    )
-    .wait()
-    .target_path.joinpath("manifest.json")
-)
-```
+If you look at the dbt project’s `/target` directory, you’ll see it stores the artifacts. When reloading your code in local development, when using `dagster_dev`, you'll see that a new manifest file is generated.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md:25]:
```suggestion
If you look at the dbt project’s `/target` directory, you’ll see it stores the artifacts. When you use `dagster dev` in local development and you reload your code, you'll see that a new manifest file is generated.
```

File: docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md [hunk 9]
```diff
@@ -6,79 +6,30 @@ lesson: '4'
...
-   ```
-
-3. Finally, let’s check to see if the variable is set:
-
-   - **If it is**, we’ll use our new logic to generate a new manifest file every time the code location is built
-   - **If it isn’t**, then we’ll use our old logic of depending on a specific `manifest.json` in the `target` directory.
+## Creating the manifest for production
 
-   Copy and paste the code to finalize the definition of `dbt_manifest_path`:
+This is great, however, it only handles the preparation of a new manifest file in local development. In production, where a dbt project is stable, we may want to prepare a new manifest file only at build time, during the deployment process. This can be done using the command line interface (CLI) available in the `dagster_dbt` package installed earlier.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-4/2-speeding-up-the-development-cycle.md:33]:
```suggestion
This is great, however, it only handles the preparation of a new manifest file in local development. In production, where a dbt project is stable, we may want to prepare a new manifest file only at build time, during the deployment process. This can be done using the command line interface (CLI) available in the `dagster_dbt` package.
```

File: docs/dagster-university/pages/dagster-dbt/lesson-7/4-creating-the-manifest-during-deployment.md [hunk 10]
```diff
@@ -6,32 +6,37 @@ lesson: '7'
...
+Building your manifest for your production deployment will be needed for both open source and Dagster+ deployments. In this case, Dagster+’s out-of-the-box `deploy.yml` GitHub Action isn’t aware that you’re also trying to deploy a dbt project with Dagster.
 
-Since your CI/CD will be running in a fresh environment, you'll need to install dbt and run `dbt deps` before building your manifest with `dbt parse`.
+Since your CI/CD will be running in a fresh environment, you'll need to install dbt and other dependencies before building your manifest.
 
-To get our deployment working, we need to add a step to our GitHub Actions workflow that runs the dbt commands required to generate the `manifest.json`. Specifically, we need to run `dbt deps` and `dbt parse` in the dbt project, just like you did during local development.
+To get our deployment working, we need to add a step to our GitHub Actions workflow that runs the commands required to generate the `manifest.json`. Specifically, we need to run the `dbt project prepare-and-package` command, available in the `dagster_dbt` package discussed earlier.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-7/4-creating-the-manifest-during-deployment.md:15]:
```suggestion
To get our deployment working, we need to add a step to our GitHub Actions workflow that runs the commands required to generate the `manifest.json`. Specifically, we need to run the `dbt project prepare-and-package` command, available in the `dagster_dbt` package.
```

Suggesting we remove this because it sounds like it refers to the CLI command, which I haven't seen until now. We also just don't need the extra words.

File: docs/dagster-university/pages/dagster-dbt/lesson-7/5-preparing-for-a-successful-run.md [hunk 11]
```diff
@@ -61,22 +61,47 @@ Because we’re still using a DuckDB-backed database, our `type` will also be `d
...
+   target=os.getenv("DBT_TARGET")
+)
+```
+
+---
+
 ## Adding a prod target to deploy.yml
 
-Next, we need to update the dbt commands in the `.github/workflows/deploy.yml` file to target the new `prod` profile. This will ensure that dbt uses the correct connection details when the GitHub Action runs as part of our Dagster+ deployment.
+After that, we need to update the dbt commands in the `.github/workflows/deploy.yml` file to target the new `prod` profile. This will ensure that dbt uses the correct connection details when the GitHub Action runs as part of our Dagster+ deployment.
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-7/5-preparing-for-a-successful-run.md:93]: Can you change this back to `Next`? The `that` here feels vague.

File: docs/dagster-university/pages/dagster-dbt/lesson-7/5-preparing-for-a-successful-run.md [hunk 12]
```diff
@@ -61,22 +61,47 @@ Because we’re still using a DuckDB-backed database, our `type` will also be `d
...
 
-Open the file, scroll to the dbt step you added, and add `-- target prod` after the `dbt parse` command. This command should be on or around line 52:
+Open the file, scroll to the environment variable sections, and set an environment variable named `DBT_TARGET` to `prod`. This command should be on or around line 12:
```
Comment (erinkcochran87) [inline docs/dagster-university/pages/dagster-dbt/lesson-7/5-preparing-for-a-successful-run.md:95]:
```suggestion
Open the file, scroll to the environment variable section, and set an environment variable named `DBT_TARGET` to `prod`. This should be on or around line 12:
```

I'm assuming there's only one section. I'm also removing `command`, as I think it's a leftover from the previous version of this page.

**Review by erinkcochran87 — COMMENTED** [review]
Thanks for doing this! A few things, but overall it's looking good.

Let me know when this is ready to go and I'll update the quiz and lesson title in the LMS.

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md (hunk 4 above)
Comment (maximearmstrong) [reply docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md]: Rephrased in [54fd52f](https://github.com/dagster-io/dagster/pull/23098/commits/54fd52f9d9ea4631d085df33a7dccb2eb41d47d6) to `The location of ... is used as a point ...`

File: docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md (hunk 3 above)
Comment (erinkcochran87) [reply docs/dagster-university/pages/dagster-dbt/lesson-3/3-representing-the-dbt-project-in-dagster.md]: This is what I see now: 

> In the `dagster_university` directory, create a new `project.py` file and add the following imports:

But later on in the lesson, `dbt.py` is still used (ex: line 31).

**Comment by maximearmstrong** [comment]
After fully testing the updated final example [here](https://github.com/dagster-io/project-dagster-university/pull/32), I've updated this PR in commit [305f9ff](https://github.com/dagster-io/dagster/pull/23098/commits/305f9ffc7c8f06a6de35318eb1f2e8f264a788d7).

**tl;dr**: the `DbtProject` object should be on a file on its own to avoid import issues during development when calling `dagster-dbt project prepare-and-package`.

cc @cmpadden @erinkcochran87 if you want to review the changes in this commit before we merge.

**Review by erinkcochran87 — COMMENTED** [review]
Back to you!

**Review by erinkcochran87 — APPROVED** [review]
